### PR TITLE
README: How to use clusterctl

### LINKED ISSUE #226: prefer to have a yaml sample to deploy the cluster within pod
/kind bug
/sig ibmcloud

**What steps did you take and what happened:**
[A clear and concise description of what the bug is.]

we have package kind and docker client in clusterctl docker image, so it's prefer to have a yaml file which can be used to deploy cluster within pods.

cc @gyliu513 @zxDiscovery @clyang82 

**What did you expect to happen:**

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Environment:**

- Cluster-api version:
- Minikube/KIND version:
- Kubernetes version: (use `kubectl version`):
- OS (e.g. from `/etc/os-release`):

### PR DESCRIPTION
**Pull request by clyang82**
Signed-off-by: Chun Lin Yang <clyang@cn.ibm.com>

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** 
Fixes #226 

**Specify your PR reviewers**:

/cc @gyliu513 @jichenjc @xunpan @hchenxa 

**Special notes for your reviewer**:

/sig ibmcloud

### CODE REVIEW

File: docs/how_to_use_clusterctl.md [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+## How to use clusterctl image in existing kubernetes cluster?
+The `clusterctl` image is designed to run independently to provision ibmcloud cluster. we have embedded the `kind` and `kubectl` into `clusterctl` image.
+
+1. Follow [this readme](https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/#cluster-creation) to generate `cluster.yaml`, `machines.yaml`, `provider-components.yaml`, and `addons.yaml`
+
+2. Create `serviceAccount` for clusterctl running
+    ```
+    kubectl apply -f examples/clusterctl/serviceaccount.yaml
+    ```
+3. Create `configmap` with `cluster.yaml`, `machines.yaml`, `provider-components.yaml`, and `addons.yaml`.
```
Comment (hchenxa) [inline docs/how_to_use_clusterctl.md]: seems you missed the clouds.yaml here

File: docs/how_to_use_clusterctl.md [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    ```
+    kubectl create configmap ibmcloud-config --from-file=./clouds.yaml --from-file=./cluster.yaml --from-file=./machines.yaml --from-file=./provider-components.yaml
+    ```
+4. Create `secret` to store the ssh private key
+    ```
+    kubectl create secret generic ibmcloud-ssh-key-secret --from-file=id_ibmcloud=/root/.ssh/id_ibmcloud
+    ```
+5. Create clusterrole and clusterrolebinding by apply the examples/clusterrole.yaml and examples/clusterrolebinding.yaml
```
Comment (hchenxa) [inline docs/how_to_use_clusterctl.md]: why we need this step?

File: examples/clusterctl/clusterrole.yaml [hunk 3]
```diff
@@ -0,0 +1,10 @@
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: ibmcloud-service-account
+  labels:
+    app: clusterctl
+rules:
+- apiGroups: [""]
+  resources: ["secrets"]
+  verbs: ["get", "list", "watch", "create"]
```
Comment (hchenxa) [inline examples/clusterctl/clusterrole.yaml]: looks like we only need the create operations?

File: docs/how_to_use_clusterctl.md (hunk 1 above)
Comment (xunpan) [reply docs/how_to_use_clusterctl.md]: we do not need `clouds.yaml` after `provider-componets.yaml` generated

File: README.md [hunk 4]
```diff
@@ -223,6 +223,10 @@ your IBM Cloud Cluster API Kubernetes cluster.
 
 Please refer to [Create a new woker node](docs/add_node.md) for further info.
 
+### How to use clusterctl image
+
+Please refer to [How to use clusterctl image](docs/how_to_use_clusterctl.md) for further info.
```
Comment (xunpan) [inline README.md]: Could you add simple introduction when someone need to read this document?
For trouble shooting and create worker nodes, it is clear from the title. But for clusterctl image, it is misunderstanding I think.

File: docs/how_to_use_clusterctl.md [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    ```
+    kubectl get secret kubeconfig -ojsonpath={.data.kubeconfig} | base64 -d > /tmp/kubeconfig
+    ```
+8. Now it is ready to access the remote cluster
+    ```
+    kubectl --kubeconfig /tmp/kubeconfig get ns
+    NAME                       STATUS   AGE
+    default                    Active   4m11s
+    ibmcloud-provider-system   Active   3m52s
+    kube-node-lease            Active   4m14s
```
Comment (xunpan) [inline docs/how_to_use_clusterctl.md]: what is this? I did not found it in my case. I guess we'd better list only default ones for easy check by end user.

File: docs/how_to_use_clusterctl.md [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+- [IBM Cloud](#ibm-cloud)
+  - [How to use clusterctl image in existing kubernetes cluster?](#how-to-use-clusterctl-image-in-existing-kubernetes-cluster)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+
+# IBM Cloud
+## How to use clusterctl image in existing kubernetes cluster?
+The `clusterctl` image is designed to run independently to provision ibmcloud cluster. we have embedded the `kind` and `kubectl` into `clusterctl` image.
+
+1. Follow [this readme](https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/#cluster-creation) to generate `cluster.yaml`, `machines.yaml`, `provider-components.yaml`, and `addons.yaml`
```
Comment (xunpan) [inline docs/how_to_use_clusterctl.md]: `addon.yaml` is optional. But it seems there is no description for this and your job yaml file make it mandatory. We should clarify this.

File: README.md [hunk 7]
```diff
@@ -223,6 +223,10 @@ your IBM Cloud Cluster API Kubernetes cluster.
 
 Please refer to [Create a new woker node](docs/add_node.md) for further info.
 
+### How to use clusterctl image
```
Comment (gyliu513) [inline README.md]: Please also run `doctoc` to update ToC.

File: examples/clusterctl/clusterrolebinding.yaml [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+  labels:
+    app: clusterctl
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: ClusterRole
+  name: ibmcloud-service-account
+subjects:
+  - kind: ServiceAccount
+    name: ibmcloud-service-account
+
```
Comment (gyliu513) [inline examples/clusterctl/clusterrolebinding.yaml]: kill this line

File: docs/how_to_use_clusterctl.md [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+
+2. Create `serviceAccount` for clusterctl running
+    ```
+    kubectl apply -f examples/clusterctl/serviceaccount.yaml
+    ```
+3. Create `configmap` with `cluster.yaml`, `machines.yaml`, `provider-components.yaml`, and `addons.yaml`.
+    ```
+    kubectl create configmap ibmcloud-config --from-file=./cluster.yaml --from-file=./machines.yaml --from-file=./provider-components.yaml --from-file=./addons.yaml
```
Comment (gyliu513) [inline docs/how_to_use_clusterctl.md]: For some long command, you can update it as follows:

```
kubectl create configmap ibmcloud-config --from-file=./cluster.yaml \
    --from-file=./machines.yaml --from-file=./provider-components.yaml \
    --from-file=./addons.yaml
```

Ditto for other long command.

File: docs/how_to_use_clusterctl.md (hunk 5 above)
Comment (clyang82) [reply docs/how_to_use_clusterctl.md]: that is example output of `kubectl --kubeconfig /tmp/kubeconfig get ns`

File: README.md (hunk 4 above)
Comment (clyang82) [reply README.md]: Done.

File: README.md (hunk 7 above)
Comment (clyang82) [reply README.md]: good catch. fixed.

File: docs/how_to_use_clusterctl.md (hunk 6 above)
Comment (clyang82) [reply docs/how_to_use_clusterctl.md]: I think `addons.yaml` is optional  and it was declared in README file.

Comment (xunpan) [reply docs/how_to_use_clusterctl.md]: For yaml file, I mean `examples/clusterctl/job.yaml`.
```       - command: ['sh', '-c', 'clusterctl create cluster --bootstrap-type kind --provider ibmcloud -c /examples/cluster.yaml -m /examples/machines.yaml -p /examples/provider-components.yaml -a /examples/addons.yaml && kubectl create secret generic kubeconfig --from-file kubeconfig']

Comment (clyang82) [reply docs/how_to_use_clusterctl.md]: I am not sure where should be a right place to add such notice to notify the user. Maybe remove the `addons.yaml` from `examples/clusterctl/job.yaml`. and tell user that need manually modify the `job` and `configmap` if you want to add `addons.yaml`. WDYT?

Comment (xunpan) [reply docs/how_to_use_clusterctl.md]: I think it is fine. Only use mandatory should be ok for most of the case. User who want to use optional ones should do it by themselves.

Comment (clyang82) [reply docs/how_to_use_clusterctl.md]: Done.

**Pull request merged**